plugins/modules/cloud/misc/proxmox_nic.py

```diff
@@ -0,0 +1,348 @@
+
+
+def get_vmid(proxmox, name):
+    return [vm['vmid'] for vm in proxmox.cluster.resources.get(type='vm') if vm.get('name') == name]
+
+
+def get_vm(proxmox, vmid):
+    return [vm for vm in proxmox.cluster.resources.get(type='vm') if vm['vmid'] == int(vmid)]
+
+
+def update_nic(module, proxmox, vmid, interface, model, **kwargs):
+    vm = get_vm(proxmox, vmid)
+
+    try:
+        vminfo = proxmox.nodes(vm[0]['node']).qemu(vmid).config.get()
+    except Exception as e:
+        module.fail_json(msg='Getting information for VM with vmid = %s failed with exception: %s' % (vmid, e))
+
+    if interface in vminfo:
+        # Convert the current config to a dictionary
+        config = vminfo[interface].split(',')
+        config.sort()
+
+        config_current = {}
+
+        for i in config:
+            kv = i.split('=')
+            try:
+                config_current[kv[0]] = kv[1]
+            except IndexError:
+                config_current[kv[0]] = ''
+
+        # determine the current model nic and mac-address
+        models = ['e1000', 'e1000-82540em', 'e1000-82544gc', 'e1000-82545em', 'i82551', 'i82557b',
+                  'i82559er', 'ne2k_isa', 'ne2k_pci', 'pcnet', 'rtl8139', 'virtio', 'vmxnet3']
+        current_model = set(models) & set(config_current.keys())
+        current_model = current_model.pop()
+        current_mac = config_current[current_model]
+
+        # build nic config string
+        config_provided = "{0}={1}".format(model, current_mac)
+    else:
+        config_provided = model
+
+    if kwargs['mac']:
+        config_provided = "{0}={1}".format(model, kwargs['mac'])
+
+    if kwargs['bridge']:
+        config_provided += ",bridge={0}".format(kwargs['bridge'])
+
+    if kwargs['firewall']:
+        config_provided += ",firewall=1"
+
+    if kwargs['link_down']:
+        config_provided += ',link_down=1'
+
+    if kwargs['mtu']:
+        if model == 'virtio':
+            config_provided += ",mtu={0}".format(kwargs['mtu'])
+        else:
+            module.fail_json(
+                vmid=vmid,
+                msg='Unable to set MTU for nic {0} on VM with vmid {1}, model should'
+                    ' be set to \'virtio\': '.format(interface, vmid))
+
+    if kwargs['queues']:
+        config_provided += ",queues={0}".format(kwargs['queues'])
+
+    if kwargs['rate']:
+        config_provided += ",rate={0}".format(kwargs['rate'])
+
+    if kwargs['tag']:
+        config_provided += ",tag={0}".format(kwargs['tag'])
+
+    if kwargs['trunks']:
+        config_provided += ",trunks={0}".format(';'.join(str(x) for x in kwargs['trunks']))
+
+    net = {interface: config_provided}
+    vm = get_vm(proxmox, vmid)
+
+    if ((interface not in vminfo) or (vminfo[interface] != config_provided)):
+        if not module.check_mode:
+            if proxmox.nodes(vm[0]['node']).qemu(vmid).config.set(**net) is None:
+                return True
+        else:
+            return True
```

Comment:
If you rewrite this as
```suggestion
        if module.check_mode:
            return True
        return proxmox.nodes(vm[0]['node']).qemu(vmid).config.set(**net) is None
```
it's easier to understand IMO.

What exactl does `set()` return other than `None` btw? If it is an error, I think you should handle that instead of just ignoring it or returning `changed == False`.

Reply:
Agreed it is more readable that way! I'll commit the suggestions.

`set()` is transparently forwarded to the PUT method on https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/qemu/{vmid}/config the only thing it returns is `None`.

Reply:
In that case, how about:
```
if not module.check_mode:
    proxmox.nodes(vm[0]['node']).qemu(vmid).config.set(**net)
return True
```
?